#Transformer-Based Sentiment Classification using DistilBERT

## Overview
This project uses a transformer-based language model to classify text sentiment. The goal is to explore how modern NLP models such as BERT can capture contextual meaning in text more effectively than traditional machine learning approaches.

## Research Question
How effectively can transformer-based models classify sentiment in text compared to traditional machine learning approaches in terms of accuracy and contextual understanding?

In [3]:
!pip install transformers datasets torch scikit-learn pandas

import pandas as pd
from datasets import load_dataset
from transformers import pipeline
from sklearn.metrics import accuracy_score

In [4]:
#Load Dataset
dataset = load_dataset("imdb")

test_data = dataset["test"].shuffle(seed=42).select(range(200))

texts = test_data["text"]
labels = test_data["label"]

df = pd.DataFrame({"text": texts, "true_label": labels})
df.head()

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md: 0.00B [00:00, ?B/s]

plain_text/train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

plain_text/test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

plain_text/unsupervised-00000-of-00001.p(…):   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

,text,true_label
0,<br /><br />When I unsuspectedly rented A Thou...,1
1,This is the latest entry in the long series of...,1
2,This movie was so frustrating. Everything seem...,0
3,"I was truly and wonderfully surprised at ""O' B...",1
4,This movie spends most of its time preaching t...,0


## Dataset
The dataset used is the IMDB movie review dataset from the Hugging Face datasets library. Reviews are labeled as either positive or negative sentiment.

For this project, a subset of the test set is used to evaluate a transformer-based sentiment classification model.

In [5]:
#Load BERT-style sentiment pipeline
classifier = pipeline("sentiment-analysis")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

In [7]:
#Generate predictions
predictions = classifier(list(df["text"][:50]), truncation=True)
predictions[:5]

[{'label': 'POSITIVE', 'score': 0.9988757967948914},
 {'label': 'POSITIVE', 'score': 0.9969831109046936},
 {'label': 'NEGATIVE', 'score': 0.9972442388534546},
 {'label': 'NEGATIVE', 'score': 0.649212658405304},
 {'label': 'NEGATIVE', 'score': 0.9985034465789795}]

In [8]:
#Convert labels
predicted_labels = [1 if pred["label"] == "POSITIVE" else 0 for pred in predictions]
true_labels = list(df["true_label"][:50])

accuracy = accuracy_score(true_labels, predicted_labels)
print("Accuracy:", accuracy)

Accuracy: 0.94


In [9]:
#Show sample predictions
sample_df = pd.DataFrame({
    "text": list(df["text"][:5]),
    "true_label": true_labels[:5],
    "predicted_label": predicted_labels[:5],
    "confidence": [round(pred["score"], 4) for pred in predictions[:5]]
})

sample_df

,text,true_label,predicted_label,confidence
0,<br /><br />When I unsuspectedly rented A Thou...,1,1,0.9989
1,This is the latest entry in the long series of...,1,1,0.9970
2,This movie was so frustrating. Everything seem...,0,0,0.9972
3,"I was truly and wonderfully surprised at ""O' B...",1,0,0.6492
4,This movie spends most of its time preaching t...,0,0,0.9985


#Results
The transformer-based model achieved high accuracy (0.94) on the evaluation subset of the IMDB dataset, demonstrating strong performance in sentiment classification tasks.

Sample predictions indicate that the model is able to correctly classify both positive and negative reviews with high confidence. The results demonstrate a clear improvement over traditional machine learning approaches used in earlier projects.

#Interpretation
The results show that transformer-based models are highly effective for sentiment classification tasks. By leveraging contextual language representations, the model can capture nuances in text that simpler models often miss.

Compared to traditional approaches such as TF-IDF with Logistic Regression, the transformer model achieves higher accuracy and provides more reliable predictions, particularly for complex or context-dependent sentences.

#Limitations
- Only a small subset of the dataset was used for evaluation
- No fine-tuning of the transformer model was performed
- The model may struggle with very long or highly ambiguous text
- Computational cost is higher compared to traditional machine learning models

# Fututre Work
- Fine-tune the transformer model on the full dataset
- Compare performance with other transformer architectures (e.g., BERT, RoBERTa)
- Analyze misclassified examples to better understand model limitations
- Apply the model to real-world datasets with more diverse text